### PR TITLE
Introduce an `ignore_above` index-level setting

### LINKED ISSUE #112795: Introduce `ignore_above` index level setting
### Description

#108720 will change the default value for `ignore_above` when LogsDB index mode is adopted. Before being able to do that we need to introduce a new index-level setting `ignore_above` that is applied to all text-like fields for a specific index mapping. The index level setting will default to `1024`. Text-like field mappings will be able to override the index level setting using `ignore_above` with the mapping-level setting taking precedence on the index-level setting.

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Here we introduce a new index-level setting, `ignore_above`, similar to what we have
for `ignore_malformed`. The setting will apply to all `keyword`, `wildcard` and `flattened`
fields. Each field mapping will still be allowed to override the index-level setting using a
mapping-level `ignore_above` value.

Resolves #112795

### CODE REVIEW

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+          fields:
+            - keyword
+            - flattened
+          query:
+            match_all: {}
+
+  - length: { hits.hits: 1 }
+  - match: { hits.hits.0._source.keyword: "foo bar" }
+  - match: { hits.hits.0._source.flattened.value: "the quick brown fox" }
+  - match: { hits.hits.0.fields.keyword.0: "foo bar" }
```
Comment (lkts) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml:40]: Should we assert that flattened is null?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        body:
+          fields:
+            - keyword
+            - flattened
+          query:
+            match_all: {}
+
+  - length: { hits.hits: 1 }
+#TODO: synthetic source field reconstruction bug (see https://github.com/elastic/elasticsearch/issues/112044)
+#  - match: { hits.hits.0._source.keyword: ["foo bar", "the quick brown fox"] }
```
Comment (lkts) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml:84]: The bug is for flattened but keyword is commented out?

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 3]
```diff
@@ -63,15 +60,20 @@
 public abstract class FieldMapper extends Mapper {
     private static final Logger logger = LogManager.getLogger(FieldMapper.class);
 
-    public static final Setting<Boolean> IGNORE_MALFORMED_SETTING = Setting.boolSetting("index.mapping.ignore_malformed", settings -> {
-        if (IndexSettings.MODE.get(settings) == IndexMode.LOGSDB
-            && IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(settings).onOrAfter(IndexVersions.ENABLE_IGNORE_MALFORMED_LOGSDB)) {
```
Comment (lkts) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:68]: Bad merge? This is not supposed to be reverted.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 4]
```diff
@@ -63,15 +60,20 @@
...
-            return "true";
-        } else {
-            return "false";
-        }
-    }, Property.IndexScope, Property.ServerlessPublic);
+    public static final Setting<Integer> IGNORE_ABOVE_SETTING = Setting.intSetting(
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:63]: Why is this added to `FieldMapper`? It only applies to a handful of mappers, iiuc.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/ValuesSourceReaderBenchmark.java [hunk 5]
```diff
@@ -183,7 +183,9 @@ private static BlockLoader blockLoader(String name) {
                 Lucene.KEYWORD_ANALYZER,
                 Lucene.KEYWORD_ANALYZER,
                 Lucene.KEYWORD_ANALYZER,
-                new KeywordFieldMapper.Builder(name, IndexVersion.current()).docValues(ft.docValuesType() != DocValuesType.NONE),
+                new KeywordFieldMapper.Builder(name, Integer.MAX_VALUE, IndexVersion.current()).docValues(
```
Comment (kkrik-es) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/ValuesSourceReaderBenchmark.java:186]: This is repeated in most tests.. maybe add a factory method `KeywordFieldMapper.builderForTests(String name)`.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+          fields:
+            - keyword
+            - flattened
+          query:
+            match_all: {}
+
+  - length: { hits.hits: 1 }
+  - match: { hits.hits.0._source.keyword: ["foo bar", "the quick brown fox"] }
+  - match: { hits.hits.0._source.flattened.value: ["the quick brown fox", "jumps over"] }
+  - match: { hits.hits.0.fields.keyword.0: "foo bar" }
```
Comment (kkrik-es) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml:81]: Assert that length is 1?

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java (hunk 4 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:63]: It is where `ignore_malformed` is...which looks a good place for me and despite `ignore_malformed` being for many more types is still not valid for all of them. Moreover `FieldMapper` is already imported by `WildcardFieldMapper` (which BTW is in xpack), and is in the same package of `KeywordFieldMapper` and `FlattenedFieldMapper`...which means also no additional import (and as such no additional dependency) is required. I don't see a better common class...the other option would be `Mapper`...but that I guess would trigger the same question. Any suggestion?

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java [hunk 7]
```diff
@@ -193,7 +187,13 @@ public static final class Builder extends FieldMapper.DimensionBuilder {
         private final ScriptCompiler scriptCompiler;
         private final IndexVersion indexCreatedVersion;
 
-        public Builder(String name, IndexAnalyzers indexAnalyzers, ScriptCompiler scriptCompiler, IndexVersion indexCreatedVersion) {
+        public Builder(
+            String name,
+            IndexAnalyzers indexAnalyzers,
+            ScriptCompiler scriptCompiler,
+            int ignoreAboveDefault,
```
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:194]: Unless there are at least a few settings to fetch I would rather keep passing a native integer instead of a reference to an object. A native integer passed on the stack is a bit easier to "clean up" when a method goes out of scope and does not require work from the garbage collector. I would agree with you if we were using more settings...but I don't see any other setting at the moment.

Also for testing purposes is just easier to pass an integer rather than a whole `Settings`.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml (hunk 6 above)
Comment (salvatore-campagna) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml:81]: If length is not 1 this `match` would fail, right and the failure would be something like

```
expected "foo bar" but was "[foo bar", "baz bax"]
```

or

```
expected "foo bar" but was null
```

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java (hunk 4 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:63]: It may be worth adding an intermediate class for such functionality, e.g. `BaseTextProcessingMapper` or something like that. Another one would be to just add it to `KeywordFieldMapper` and reference it elsewhere, since most other mappers use it. Not ideal, but seems preferable to `FieldMapper` imho.

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:194]: Also this builder is also used by other classes making use of `KeywordFieldMapper` like `PercolatorFieldMapper`. Passing `Setting` here would require changes to propagate settings there which seems unnecessary to me.

[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:194]: But then, you need to reference the index param you want to fetch in all interesting call sites - somewhat leaking an internal detail.. I'd also focus on the consistency argument, let's follow existing patterns where possible.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml (hunk 2 above)
Comment (salvatore-campagna) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml:84]: No the bug is for `keyword` I just forgot to create an issue similar to the flattened one. I put the link in the comment for my own reference since I believe it might be the same issue affecting both `keyword` and `flattened`). I would like to create a reproduction which uses the mapping level `ignore_above` instead of the setting level `ignore_above`.

Comment (salvatore-campagna) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml:84]: I am also a bit surprised that I don't see it for the flattened field too...but I didn't investigate that.

Comment (salvatore-campagna) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml:84]: I am going to remove the link and try to reproduce the issue with another test and link it to the flattened one.

Comment (salvatore-campagna) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml]: At the moment I am not able to reproduce the issue with `keyword` with just a `keyword` field. This makes me believe the actual issue is with flattened fields.

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:194]: The fact that is done in `NumericFieldMapper` does not meat it is consistent...

See: LegacyGeoShapeFieldMapper, BooleanFieldMapper, DateFieldMapper,GeoPointFieldMapper...

If we want to be consistent maybe they should all be like that. My PR touches like 34 files already....I would prefer not to make it like 50 or 70. This "consistency" refactoring could be done in a separate one.

[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:200]: Note also that `NumberFieldMapper` just defines an additional constructor that takes a `Settings` and calls

```
public Builder(
            String name,
            NumberType type,
            ScriptCompiler compiler,
            boolean ignoreMalformedByDefault,
            boolean coerceByDefault,
            IndexVersion indexCreatedVersion,
            IndexMode mode
        )
```

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml (hunk 6 above)
Comment (salvatore-campagna) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/530_ignore_above_stored_source.yml:82]: Moreover if I assert the length before asserting the match a test would fail with something like

```
expected length 1 but was 2 m(or whatever)
```

while I think is a bit better to actually see the field values in error message. We would need to re-run the test removing the length assertion to see which value is returned.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java (hunk 4 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:66]: I'm still against polluting `FieldMapper` for functionality belonging to a small subset of subclasses. You can ask @martijnvg to approve this.

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:200]: Moreover, if I replace this call in the `PercolatorFieldMapper`

```
KeywordFieldMapper.Builder queryMetadataFieldBuilder = new KeywordFieldMapper.Builder(
                name,
                Integer.MAX_VALUE,
                indexCreatedVersion
            );
```

with something like

```
KeywordFieldMapper.Builder queryMetadataFieldBuilder = new KeywordFieldMapper.Builder(
                name,
                Settings.builder().build(),
                indexCreatedVersion
            );
```

I would need to do a `getAsInt(String setting, Integer defaultValue)` with `Integer.MAX_VALUE` as `defaultValue`. This is bug-prone...beucase if instead of empty settings we pass (by mistake) settings including `index.mapping.ingore_above` then this parameter would be sued by the percolator too.

[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:200]: What about adding a constructor to the `KeywordFieldMapper.Builder` class that just accepts a `MappingParserContext` as parameter? In production code, all current parameter s(except name) come from it.

 This new constructor would delegate to the existing constructor with 4 parameters (we should change its visibility to private, so that constructor with `String, IndexVersion` can still be used). 

Then next time we add a setting we most likely don't need to change the constructor, which results in a much smaller change.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java (hunk 4 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:66]: Ignore malformed is a different feature than ignore above. Te ignore above feature only exists for a few field mappers.  We don't have a common base class for these field mappers. I think we should just move the setting to  `KeywordFieldMapper`.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:66]: Or maybe better just move the setting to `IndexSettings`?

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java [hunk 8]
```diff
@@ -143,6 +142,10 @@ private static KeywordFieldMapper toType(FieldMapper in) {
         return (KeywordFieldMapper) in;
     }
 
+    public static Builder buildForTest(final String name, boolean hasDocValues) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:145]: I don't think this is needed and is unrelated to adding the new index setting? We can keep using the constructor  with two parameters?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+      search:
+        body:
+          fields:
+            - keyword
+            - flattened
+          query:
+            match_all: {}
+
+  - length: { hits.hits: 1 }
+  #TODO: synthetic source field reconstruction bug (TBD: add link to the issue here)
```
Comment (salvatore-campagna) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/540_ignore_above_synthetic_source.yml:83]: I could not reproduce the issue using only a `keyword` field...it looks like the issue happens when a keyword is used together with a flattened field.

**Comment by bitomukesh** [comment]
<div>
<h2>Changelist by Bito</h2>
    <p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>New Feature</b> - Introduce index-level ignore_above setting
</td>
<td width="70%">


<div>

-&nbsp;[`ignore-above.asciidoc`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-ad81490a6d2f1112c31e8308ca1349a7be50ee839345f86b4e5be4f4d8c6e845) - Added documentation for the new index-level ignore_above setting
<div>

-&nbsp;[`530_ignore_above_stored_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-d01623834cd995c01f5595d13b67e64dbf56e77831b64c83bf5d6cdd137f336c) - Added tests for the new index-level ignore_above setting with stored _source
<div>

-&nbsp;[`540_ignore_above_synthetic_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-5e3cb051d815f94ae6f2f5739e93ed363657afcec37687499c708fc53f90a0a9) - Added tests for the new index-level ignore_above setting with synthetic _source
<div>

-&nbsp;[`550_ignore_above_invalid.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-ec6010613bd6b6cfaa7953ab530d34d723f954222765fb728f3cccf5b9ee731b) - Added tests for invalid ignore_above values
<div>

-&nbsp;[`IndexScopedSettings.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-e22a7734dca6b97f6c89a7d888955054d0512def7a561f465d1a4e85279d4e46) - Added ignore_above setting to index-scoped settings
<div>

-&nbsp;[`IndexSettings.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-aefe21240b812c83c886f5b30e4e849fbb9aa68a46eec19930844db34b58954e) - Implemented the new index-level ignore_above setting
<div>

-&nbsp;[`KeywordFieldMapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-eb2239fb7344f4f8d868307ec66a877fedec6ad509644678580793be67b2e0c3) - Updated KeywordFieldMapper to use the new index-level ignore_above setting
<div>

-&nbsp;[`Mapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-0fb30508939be757a0fdda0f93c0428cb436a2762a0a2645b3400de498b2bb3d) - Added a new node feature for the ignore_above index-level setting
<div>

-&nbsp;[`MapperFeatures.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-c8ad37fc6b7723f47d855544d8ebff68e2057f73aa792a62f56e9b7511a3ec4f) - Added the new ignore_above index-level setting feature
<div>

-&nbsp;[`FlattenedFieldMapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-54e819b3829ea46dafc9673fddb759908a811d5cc02e5f5c6f84927c8ee7fd8f) - Updated FlattenedFieldMapper to use the new index-level ignore_above setting
<div>

-&nbsp;[`KeywordFieldTypeTests.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-a37f0f63173f2299ae23cffa321274fbc42efc94a7a7ea67896ddb06e0032c57) - Updated tests for KeywordFieldType to include the new ignore_above setting
<div>

-&nbsp;[`MultiFieldsTests.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-e41dd57dc2c147fd866212c1c3c85be8a0b685d47d688f530dd8c4d1a3a9c0e3) - Updated MultiFieldsTests to include the new ignore_above setting
<div>

-&nbsp;[`20_ignore_above_stored_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-7c9ba9c57fea03cd43b6236fe71a3cce28c35552046282e0091bed937fb375db) - Added tests for wildcard field type with the new ignore_above setting
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>New Feature</b> - Introduce index-level ignore_above setting
</td>
<td width="70%">


<div>

-&nbsp;[`30_ignore_above_synthetic_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-595bdc4da4315b813a8a10b441ef3327a09593a65ecf9c6bcf17f6b2ba94adfe) - Added new test for wildcard field type with ignore_above setting and synthetic source
<div>

-&nbsp;[`WildcardFieldMapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-d3c41c933d0a18037375ed593ed93fca61dd4e244ef299b0ad57f6a7d1cdabda) - Updated WildcardFieldMapper to support the new index-level ignore_above setting
<div>

-&nbsp;[`CoreTestTranslater.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-2523e7e543ff4f9f20d006cd600b5f41f8c003af7019d19438674eb20c03ac5c) - Added logic to handle ignore_above setting in runtime fields tests
<div>

-&nbsp;[`ignore-above.asciidoc`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-ad81490a6d2f1112c31e8308ca1349a7be50ee839345f86b4e5be4f4d8c6e845) - Added documentation for the new index-level ignore_above setting
<div>

-&nbsp;[`530_ignore_above_stored_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-d01623834cd995c01f5595d13b67e64dbf56e77831b64c83bf5d6cdd137f336c) - Added tests for the new index-level ignore_above setting with stored _source
<div>

-&nbsp;[`540_ignore_above_synthetic_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-5e3cb051d815f94ae6f2f5739e93ed363657afcec37687499c708fc53f90a0a9) - Added tests for the new index-level ignore_above setting with synthetic _source
<div>

-&nbsp;[`550_ignore_above_invalid.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-ec6010613bd6b6cfaa7953ab530d34d723f954222765fb728f3cccf5b9ee731b) - Added tests for invalid ignore_above values
<div>

-&nbsp;[`IndexScopedSettings.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-e22a7734dca6b97f6c89a7d888955054d0512def7a561f465d1a4e85279d4e46) - Added ignore_above setting to index-scoped settings
<div>

-&nbsp;[`IndexSettings.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-aefe21240b812c83c886f5b30e4e849fbb9aa68a46eec19930844db34b58954e) - Implemented the new index-level ignore_above setting
<div>

-&nbsp;[`KeywordFieldMapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-eb2239fb7344f4f8d868307ec66a877fedec6ad509644678580793be67b2e0c3) - Updated KeywordFieldMapper to use the new index-level ignore_above setting
<div>

-&nbsp;[`Mapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-0fb30508939be757a0fdda0f93c0428cb436a2762a0a2645b3400de498b2bb3d) - Added a new node feature for the ignore_above index-level setting
<div>

-&nbsp;[`MapperFeatures.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-c8ad37fc6b7723f47d855544d8ebff68e2057f73aa792a62f56e9b7511a3ec4f) - Added the new ignore_above index-level setting feature
<div>

-&nbsp;[`FlattenedFieldMapper.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-54e819b3829ea46dafc9673fddb759908a811d5cc02e5f5c6f84927c8ee7fd8f) - Updated FlattenedFieldMapper to use the new index-level ignore_above setting
<div>

-&nbsp;[`KeywordFieldTypeTests.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-a37f0f63173f2299ae23cffa321274fbc42efc94a7a7ea67896ddb06e0032c57) - Updated tests for KeywordFieldType to include the new ignore_above setting
<div>

-&nbsp;[`MultiFieldsTests.java`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-e41dd57dc2c147fd866212c1c3c85be8a0b685d47d688f530dd8c4d1a3a9c0e3) - Updated MultiFieldsTests to include the new ignore_above setting
<div>

-&nbsp;[`20_ignore_above_stored_source.yml`](https://github.com/elastic/elasticsearch/pull/113121/files#diff-7c9ba9c57fea03cd43b6236fe71a3cce28c35552046282e0091bed937fb375db) - Added tests for wildcard field type with the new ignore_above setting
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 10]
```diff
@@ -29,6 +29,7 @@
 public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
     public static final NodeFeature SYNTHETIC_SOURCE_KEEP_FEATURE = new NodeFeature("mapper.synthetic_source_keep");
+    public static final NodeFeature IGNORE_ABOVE_INDEX_LEVEL_SETTING = new NodeFeature("mapper.ignore_above_index_level_setting");
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:32]: maybe move this constant also to `IndexSettings`?

File: x-pack/qa/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/test/CoreTestTranslater.java [hunk 11]
```diff
@@ -222,10 +222,32 @@ public boolean modifySections(List<ExecutableSection> executables) {
          */
         protected abstract boolean modifySearch(ApiCallSection search);
 
+        private static Object getSetting(final Object map, final String... keys) {
```
Comment (martijnvg) [inline x-pack/qa/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/test/CoreTestTranslater.java:225]: Why is this change needed?

**Review by martijnvg — APPROVED** [review]
Left two comments, otherwise LGTM.

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java (hunk 8 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java]: I removed this after changing the available constructors.

File: x-pack/qa/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/test/CoreTestTranslater.java (hunk 11 above)
Comment (salvatore-campagna) [reply x-pack/qa/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/test/CoreTestTranslater.java:225]: This is required to skip testing using runtime fields when `ignore_above` is used. This is done already when `ignroe_above` is used at mapping level (runtime fields do not support `ignroe_above`). 
See this code below:
```
if (propertyMap.containsKey("ignore_above")) {
                    // Scripts don't support ignore_above so we skip those fields
                    continue;
                }
```

Now, since the `ignore_above` is at index-level is is no longer at field mapping level and so tests are not skipped.

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/113414)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
